File: starlane_mobile/starlane_client/lib/data/models/service.dart
// Path: starlane_mobile/starlane_client/lib/data/models/service.dart
import 'package:equatable/equatable.dart';
import 'package:json_annotation/json_annotation.dart';

part 'service.g.dart';

@JsonSerializable()
class Service extends Equatable {
  @JsonKey(name: '_id')
  final String id;
  final String name;
  final String description;
  final String? shortDescription;
  final ServiceCategory category;
  final String serviceType;
  final ServicePricing pricing;
  final String icon;
  final List<ServiceImage> images;
  final List<String> features;
  final List<String> included;
  final bool isActive;
  final bool isFeatured;
  final int priority;
  final ServiceDetails? serviceDetails;
  final ServiceStats stats;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Service({
    required this.id,
    required this.name,
    required this.description,
    this.shortDescription,
    required this.category,
    required this.serviceType,
    required this.pricing,
    required this.icon,
    this.images = const [],
    this.features = const [],
    this.included = const [],
    this.isActive = true,
    this.isFeatured = false,
    this.priority = 0,
    this.serviceDetails,
    required this.stats,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Service.fromJson(Map<String, dynamic> json) => _$ServiceFromJson(json);
  Map<String, dynamic> toJson() => _$ServiceToJson(this);

  // Getters utilitaires
  String get displayDescription => shortDescription ?? description;
  
  String get formattedPrice {
    if (pricing.isCustomQuote) {
      return 'Sur devis';
    }
    
    const symbols = {'EUR': '€', 'USD': '\$', 'GBP': '£'};
    final symbol = symbols[pricing.currency] ?? pricing.currency;
    
    String suffix = '';
    switch(pricing.priceType) {
      case PriceType.perHour:
        suffix = '/h';
        break;
      case PriceType.perDay:
        suffix = '/jour';
        break;
      case PriceType.fixed:
        suffix = '';
        break;
      case PriceType.custom:
        suffix = '';
        break;
    }
    
    return '${pricing.basePrice.toStringAsFixed(0)}$symbol$suffix';
  }

  @override
  List<Object?> get props => [
    id, name, description, shortDescription, category, serviceType,
    pricing, icon, images, features, included, isActive, isFeatured,
    priority, serviceDetails, stats, createdAt, updatedAt
  ];
}

@JsonEnum()
enum ServiceCategory {
  @JsonValue('airTravel')
  airTravel,
  @JsonValue('transport')
  transport,
  @JsonValue('realEstate')
  realEstate,
  @JsonValue('corporate')
  corporate;

  String get displayName {
    switch (this) {
      case ServiceCategory.airTravel:
        return 'Aviation Privée';
      case ServiceCategory.transport:
        return 'Transport';
      case ServiceCategory.realEstate:
        return 'Immobilier';
      case ServiceCategory.corporate:
        return 'Corporate';
    }
  }

  String get description {
    switch (this) {
      case ServiceCategory.airTravel:
        return 'Services aériens de luxe';
      case ServiceCategory.transport:
        return 'Transport premium';
      case ServiceCategory.realEstate:
        return 'Services immobiliers';
      case ServiceCategory.corporate:
        return 'Solutions entreprise';
    }
  }

  String get iconName {
    switch (this) {
      case ServiceCategory.airTravel:
        return 'flight';
      case ServiceCategory.transport:
        return 'directions_car';
      case ServiceCategory.realEstate:
        return 'home';
      case ServiceCategory.corporate:
        return 'business';
    }
  }
}

@JsonEnum()
enum PriceType {
  @JsonValue('fixed')
  fixed,
  @JsonValue('per_hour')
  perHour,
  @JsonValue('per_day')
  perDay,
  @JsonValue('custom')
  custom;
}

@JsonSerializable()
class ServicePricing extends Equatable {
  final double basePrice;
  final String currency;
  final PriceType priceType;
  final bool isCustomQuote;

  const ServicePricing({
    required this.basePrice,
    this.currency = 'EUR',
    this.priceType = PriceType.fixed,
    this.isCustomQuote = false,
  });

  factory ServicePricing.fromJson(Map<String, dynamic> json) => 
      _$ServicePricingFromJson(json);
  Map<String, dynamic> toJson() => _$ServicePricingToJson(this);

  @override
  List<Object?> get props => [basePrice, currency, priceType, isCustomQuote];
}

@JsonSerializable()
class ServiceImage extends Equatable {
  final String url;
  final String? alt;
  final bool isPrimary;

  const ServiceImage({
    required this.url,
    this.alt,
    this.isPrimary = false,
  });

  factory ServiceImage.fromJson(Map<String, dynamic> json) => 
      _$ServiceImageFromJson(json);
  Map<String, dynamic> toJson() => _$ServiceImageToJson(this);

  @override
  List<Object?> get props => [url, alt, isPrimary];
}

@JsonSerializable()
class ServiceDetails extends Equatable {
  final String? airTravelType;
  final String? transportType;
  final String? realEstateType;
  final String? corporateType;

  const ServiceDetails({
    this.airTravelType,
    this.transportType,
    this.realEstateType,
    this.corporateType,
  });

  factory ServiceDetails.fromJson(Map<String, dynamic> json) => 
      _$ServiceDetailsFromJson(json);
  Map<String, dynamic> toJson() => _$ServiceDetailsToJson(this);

  @override
  List<Object?> get props => [
    airTravelType, transportType, realEstateType, corporateType
  ];
}

@JsonSerializable()
class ServiceStats extends Equatable {
  final int orders;
  final double revenue;

  const ServiceStats({
    this.orders = 0,
    this.revenue = 0.0,
  });

  factory ServiceStats.fromJson(Map<String, dynamic> json) => 
      _$ServiceStatsFromJson(json);
  Map<String, dynamic> toJson() => _$ServiceStatsToJson(this);

  @override
  List<Object?> get props => [orders, revenue];
}

// Request Models
@JsonSerializable()
class CreateServiceRequest extends Equatable {
  final String name;
  final String description;
  final String? shortDescription;
  final ServiceCategory category;
  final String serviceType;
  final ServicePricing pricing;
  final String icon;
  final List<ServiceImage> images;
  final List<String> features;
  final List<String> included;
  final bool isFeatured;
  final int priority;
  final ServiceDetails? serviceDetails;

  const CreateServiceRequest({
    required this.name,
    required this.description,
    this.shortDescription,
    required this.category,
    required this.serviceType,
    required this.pricing,
    required this.icon,
    this.images = const [],
    this.features = const [],
    this.included = const [],
    this.isFeatured = false,
    this.priority = 0,
    this.serviceDetails,
  });

  factory CreateServiceRequest.fromJson(Map<String, dynamic> json) => 
      _$CreateServiceRequestFromJson(json);
  Map<String, dynamic> toJson() => _$CreateServiceRequestToJson(this);

  @override
  List<Object?> get props => [
    name, description, shortDescription, category, serviceType,
    pricing, icon, images, features, included, isFeatured, priority, serviceDetails
  ];
}

@JsonSerializable()
class UpdateServiceRequest extends Equatable {
  final String? name;
  final String? description;
  final String? shortDescription;
  final ServiceCategory? category;
  final String? serviceType;
  final ServicePricing? pricing;
  final String? icon;
  final List<ServiceImage>? images;
  final List<String>? features;
  final List<String>? included;
  final bool? isFeatured;
  final int? priority;
  final ServiceDetails? serviceDetails;

  const UpdateServiceRequest({
    this.name,
    this.description,
    this.shortDescription,
    this.category,
    this.serviceType,
    this.pricing,
    this.icon,
    this.images,
    this.features,
    this.included,
    this.isFeatured,
    this.priority,
    this.serviceDetails,
  });

  factory UpdateServiceRequest.fromJson(Map<String, dynamic> json) => 
      _$UpdateServiceRequestFromJson(json);
  Map<String, dynamic> toJson() => _$UpdateServiceRequestToJson(this);

  @override
  List<Object?> get props => [
    name, description, shortDescription, category, serviceType,
    pricing, icon, images, features, included, isFeatured, priority, serviceDetails
  ];
}

File: backend/src/models/Service.js
const mongoose = require('mongoose');

const serviceSchema = new mongoose.Schema({
  // Basic Info
  name: {
    type: String,
    required: [true, 'Service name is required'],
    trim: true,
    maxlength: [100, 'Service name cannot be more than 100 characters']
  },
  description: {
    type: String,
    required: [true, 'Service description is required'],
    trim: true,
    maxlength: [500, 'Description cannot be more than 500 characters']
  },
  shortDescription: {
    type: String,
    trim: true,
    maxlength: [150, 'Short description cannot be more than 150 characters']
  },
  
  // Category
  category: {
    type: String,
    enum: {
      values: ['airTravel', 'transport', 'realEstate', 'corporate'],
      message: 'Invalid service category'
    },
    required: [true, 'Service category is required']
  },
  
  // Service Details
  serviceType: {
    type: String,
    required: [true, 'Service type is required'],
    trim: true
  },
  
  // Pricing
  pricing: {
    basePrice: {
      type: Number,
      required: [true, 'Base price is required'],
      min: [0, 'Price cannot be negative']
    },
    currency: {
      type: String,
      default: 'EUR',
      enum: ['EUR', 'USD', 'GBP']
    },
    priceType: {
      type: String,
      enum: ['fixed', 'per_hour', 'per_day', 'custom'],
      default: 'fixed'
    },
    isCustomQuote: {
      type: Boolean,
      default: false
    }
  },
  
  // Media
  icon: {
    type: String,
    required: [true, 'Service icon is required']
  },
  images: [{
    url: {
      type: String,
      required: true
    },
    alt: String,
    isPrimary: {
      type: Boolean,
      default: false
    }
  }],
  
  // Features & Details
  features: [{
    type: String,
    trim: true
  }],
  included: [{
    type: String,
    trim: true
  }],
  
  // Status & Display
  isActive: {
    type: Boolean,
    default: true
  },
  isFeatured: {
    type: Boolean,
    default: false
  },
  priority: {
    type: Number,
    default: 0
  },
  
  // Service Specific Fields
  serviceDetails: {
    // Pour Air Travel
    airTravelType: {
      type: String,
      enum: ['meetGreet', 'jetReservation', 'lounge']
    },
    
    // Pour Transport  
    transportType: {
      type: String,
      enum: ['miseDisposition', 'chauffeurVip']
    },
    
    // Pour Real Estate
    realEstateType: {
      type: String,
      enum: ['achat', 'location']
    },
    
    // Pour Corporate
    corporateType: {
      type: String,
      enum: ['evenement']
    }
  },
  
  // Statistics
  stats: {
    orders: {
      type: Number,
      default: 0
    },
    revenue: {
      type: Number,
      default: 0
    }
  }
}, {
  timestamps: true,
  toJSON: { virtuals: true },
  toObject: { virtuals: true }
});

// Indexes
serviceSchema.index({ category: 1, isActive: 1 });
serviceSchema.index({ isFeatured: 1, priority: -1 });
serviceSchema.index({ createdAt: -1 });

// Virtual pour le prix formaté
serviceSchema.virtual('formattedPrice').get(function() {
  const { basePrice, currency, priceType, isCustomQuote } = this.pricing;
  
  if (isCustomQuote) {
    return 'Sur devis';
  }
  
  const symbols = { EUR: '€', USD: '$', GBP: '£' };
  const symbol = symbols[currency] || currency;
  
  let suffix = '';
  switch(priceType) {
    case 'per_hour': suffix = '/h'; break;
    case 'per_day': suffix = '/jour'; break;
    case 'fixed': suffix = ''; break;
    default: suffix = '';
  }
  
  return `${basePrice}${symbol}${suffix}`;
});

// Static method to get featured services
serviceSchema.statics.getFeatured = function() {
  return this.find({ 
    isFeatured: true, 
    isActive: true 
  }).sort({ priority: -1, createdAt: -1 });
};

// Static method to get by category
serviceSchema.statics.getByCategory = function(category) {
  return this.find({ 
    category: category, 
    isActive: true 
  }).sort({ priority: -1, name: 1 });
};

module.exports = mongoose.model('Service', serviceSchema);